File: TMB/SIRest1.R
#rm(list=ls())
#gc(verb=TRUE)
eps = 1e-8
source('./SIR_read_dat.R')
print(" ",quote=FALSE)
print("##############################################",quote=FALSE)
SIR.dat = read.dat.file()
print(names(SIR.dat))
data = SIR.dat$data
print("----data:")
print(data)
# from python simulator
# def SIRsim(mu = 0.001, beta = 0.5, gamma = 0.075, N0 = 1e5, sigma_eye = 0.3,

init = list(
    sigma_P = 0.5, #1e3,
    sigma_beta = 0.01, #2.0,
    mu = 0.05,
    gamma = 0.075,
    sigma_C = 0.25,
    sigma_D = 0.25,
    beta = 0.2
)
print("-initial values:")
print(init)

parameters = list(
    logsigma_P = log(init$sigma_P),
    logsigma_beta = log(init$sigma_beta),
    logmu    = log(init$mu),
    loggamma = log(init$gamma),
    logsigma_C = log(init$sigma_C),
    logsigma_D = log(init$sigma_D),
    logbeta = rep(log(init$beta),data$ntime)
)
print(paste("--parameters: ", length(parameters)))
print(parameters)
#print(names(parameters))
#print(names(map))
#print(paste("map: ",length(map)))
map = list("logsigma_C" = as.factor(NA),
           "logsigma_D" = as.factor(NA))
#          "logsigma_beta" = as.factor(NA),
#          "loggamma"  = as.factor(NA),
#          "logmu"  = as.factor(NA),
#map = list(
#      logsigma_P = as.factor(1),
#      logsigma_beta = as.factor(1),
#      logmu    = as.factor(NA),
#      loggamma = as.factor(NA),
#      logsigma_C = as.factor(NA),
#      logsigma_D = as.factor(NA),
#      logbeta = as.factor(rep(1,length(parameters$logbeta)))
#)
#map = list(
#      "logsigma_P" = as.factor(1),
#      "logsigma_beta" = as.factor(1),
#      "logmu"    = as.factor(NA),
#      "loggamma" = as.factor(NA),
#      "logsigma_C" = as.factor(NA),
#      "logsigma_D" = as.factor(NA),
#      "logbeta" = as.factor(rep(1,length(parameters$logbeta)))
#)
print(paste("map: ",length(map)))
print(map)

require(TMB)

print("\nCompiling-------------------------------------",quote=FALSE)
compile("SIRest1.cpp")
dyn.load(dynlib("SIRest1"))
print("\nFinished compilation and dyn.load-------------",quote=FALSE)
print("\nCalling MakeADFun-----------------------------",quote=FALSE)
##mody=MakeADFun(data, parsy, random=c("rei", "rey"), DLL="REiy", map=mapy)
obj = MakeADFun(data,parameters,random=c("logbeta"),map=map,DLL="SIRest1")
#obj = MakeADFun(data,parameters,map=map,DLL="SIRest1")
print("\n--------MakeADFun Finished--------------------",quote=FALSE)
print(paste("par: ",obj$par))

print("\nStarting minimization-------------------------",quote=FALSE)
obj$control=list(trace=1,eval.max=1,iter.max=2)
opt <- nlminb(obj$par,obj$fn,obj$gr,control=obj$control)
##opty= nlminb(mody $par, mody $fn, mody $gr)
print("\nDone minimization-----------------------------",quote=FALSE)
print(paste("Objective function value =",opt$objective))
print(paste("Number of parameters = ",length(opt$par)),quote=FALSE)
print("parameters:\n",quote=FALSE)
print(exp(opt$par))

plot.state(data,obj)

File: TMB/SIRest1.cpp
#include <TMB.hpp>
#include <math.h>
#include "trace.h"
const double TWO_M_PI = 2.0*M_PI;
const double LOG_M_PI = log(M_PI);
//const double logZeroCatch = 0.0;  // log(1.0)
const double eps = 1e-8;

template < class Type > Type square(Type x)
{
    return x * x;
}

// zero-inflated log-normal error
template <class Type>
Type ZILNerr(Type obs, Type pred, Type var, const double prop0 = 0.15)
{
    Type tmp;

    if (obs > 0.0)
    {
        tmp = (1.0-prop0)*0.5*(log(TWO_M_PI*var) + square(log(obs) - log(pred+eps))/var);
    }
    else
    {
        tmp = prop0*0.5*(log(TWO_M_PI*var));
    }


    return tmp;
}

// log-normal error
template <class Type>
Type NLerr(Type obs, Type pred, Type var)
{
    Type nll = 0.5*(log(TWO_M_PI*var) + square(log(obs)-log(pred))/var);
    return nll;
}


template<class Type> 
Type objective_function <Type>::operator()()
{
    DATA_SCALAR(N0)
    DATA_INTEGER(ntime)
    DATA_VECTOR(obs_cases)
    DATA_VECTOR(obs_deaths)
    DATA_SCALAR(prop_zero_deaths)

    PARAMETER(logsigma_P);          // SIR process error
    PARAMETER(logsigma_beta);       // beta random walk sd
    PARAMETER(logmu);               // mortality rate of infection population
    PARAMETER(loggamma);            // recovery rate of infection population
    PARAMETER(logsigma_C);          // cases observation error
    PARAMETER(logsigma_D);          // deaths observation error
    PARAMETER_VECTOR(logbeta);      // infection rate time series

    // state variables
    vector <Type> N(ntime);         // total population
    vector <Type> Eye(ntime);       // number of infections
    vector <Type> S(ntime);         // susceptible population
    vector <Type> R(ntime);         // number of recovered
    vector <Type> D(ntime);         // number of deaths from infected population

    Type mu = exp(logmu);
    Type gamma = exp(loggamma);
    Type sigma_P = exp(logsigma_P);
    Type sigma_C = exp(logsigma_C);
    Type sigma_D = exp(logsigma_D);
    Type sigma_beta = exp(logsigma_beta);

    Type varlogbeta = square(log(sigma_beta));
    Type varlogPop = square(log(sigma_P));
    Type varlogcases = square(log(sigma_C));
    Type varlogdeaths = square(log(sigma_D));

    //  initialize variables and parameters
    N(0) = N0;
    Eye(0) = 1.0;
    S(0) = N(0) - 1.0;

    Type f = 0.0;

    //  loop over time
    for (int t = 1; t <  ntime; t++)
    {
         // infection rate random walk

         Type betanll = 0.0;
         betanll += 0.5*(log(TWO_M_PI*varlogbeta) + square(logbeta(t-1) -logbeta(t))/varlogbeta);
     
         Type Pnll = 0.0;
         N(t) = S(t-1) + Eye(t-1) + R(t-1);

         Type bison = exp(logbeta(t-1))*Eye(t-1)/(N(t-1)+eps);
         Type gEye = gamma*Eye(t-1);
         Type mEye = mu*Eye(t-1);

         Eye(t) = Eye(t-1) + ((exp(logbeta(t-1))*Eye(t-1)*S(t-1)/N(t-1)) - mu*Eye(t-1));
         Pnll += NLerr(Eye(t-1)+eps, Eye(t)+eps,varlogPop);

         S(t) = S(t-1) + exp(logbeta(t-1))*Eye(t-1)*S(t-1)/N(t-1);
         Pnll += NLerr(S(t-1)+eps, S(t)+eps,varlogPop);

         R(t) = R(t-1) + gEye;
         Pnll += NLerr(R(t-1)+eps, R(t)+eps,varlogPop);

         D(t) = mEye;

         f += (betanll + Pnll);
     }
 
     // compute observation likelihoods
     Type cnll = 0.0;
     Type dnll = 0.0;
     for (int t = 0; t < ntime; t++)
     {   
         dnll += ZILNerr(obs_deaths(t),D(t),varlogdeaths);

         cnll += NLerr(obs_cases(t),Eye(t)+eps,varlogcases);
     }

     f += (cnll + dnll);

     REPORT(S)
     REPORT(Eye)
     REPORT(R)
     REPORT(D)
     REPORT(N)
     REPORT(logbeta)
     return f;
}
